File: src/nodetool/models/asset.py
from typing import Dict, List, Optional, Literal, Sequence
from datetime import datetime
from nodetool.common.content_types import CONTENT_TYPE_TO_EXTENSION
from nodetool.common.environment import Environment

from nodetool.models.base_model import DBModel, DBField, create_time_ordered_uuid
from nodetool.models.condition_builder import ConditionBuilder, Field

log = Environment.get_logger()


# Asset Pydantic Model
class Asset(DBModel):
    @classmethod
    def get_table_schema(cls):
        return {"table_name": "nodetool_assets"}

    type: Literal["asset"] = "asset"
    id: str = DBField()
    user_id: str = DBField(default="")
    workflow_id: str | None = DBField(default=None)
    parent_id: str = DBField(default="")
    file_id: str | None = DBField(default="")
    name: str = DBField(default="")
    content_type: str = DBField(default="")
    metadata: dict | None = DBField(default=None)
    created_at: datetime = DBField(default_factory=datetime.now)
    duration: Optional[float] = DBField(default=None)

    @property
    def file_extension(self) -> str:
        """
        Get the file extension of the asset.

        For example, if the content type is "image/jpeg", this will return "jpeg".
        """
        return (
            CONTENT_TYPE_TO_EXTENSION[self.content_type]  # type: ignore
            if self.content_type in CONTENT_TYPE_TO_EXTENSION
            else "bin"
        )

    @property
    def has_thumbnail(self) -> bool:
        """
        Returns True if the asset type supports thumbnails.
        """
        return self.content_type.startswith("image/") or self.content_type.startswith(
            "video/"
        )

    @property
    def file_name(self) -> str:
        """
        Get the file name of the asset.
        """
        return f"{self.id}.{self.file_extension}"

    @property
    def thumb_file_name(self) -> str:
        """
        Get the file name of the thumbnail.
        """
        return f"{self.id}_thumb.jpg"

    @classmethod
    def create(
        cls,
        user_id: str,
        name: str,
        content_type: str,
        metadata: dict | None = None,
        parent_id: str | None = None,
        workflow_id: str | None = None,
        duration: float | None = None,
        **kwargs,
    ):
        return super().create(
            id=create_time_ordered_uuid(),
            name=name,
            user_id=user_id,
            parent_id=parent_id or user_id,
            workflow_id=workflow_id,
            content_type=content_type,
            duration=duration,
            created_at=datetime.now(),
            metadata=metadata,
            **kwargs,
        )

    @classmethod
    def find(cls, user_id: str, asset_id: str):
        """
        Find an asset by user_id and asset_id.
        """
        item = cls.get(asset_id)
        if item and item.user_id == user_id:
            return item
        return None

    @classmethod
    def paginate(
        cls,
        user_id: str,
        parent_id: Optional[str] = None,
        workflow_id: Optional[str] = None,
        content_type: Optional[str] = None,
        limit: int = 100,
        start_key: str | None = None,
        reverse: bool = False,
    ):
        """
        Paginate assets for a user using boto3.
        Applies filters for parent_id if provided.
        Returns a tuple of a list of Assets and the last evaluated key for pagination.
        Last key is "" if there are no more items to be returned.
        """
        if parent_id:
            condition = (
                Field("user_id")
                .equals(user_id)
                .and_(Field("parent_id").equals(parent_id))
                .and_(Field("id").greater_than(start_key or ""))
            )
            return cls.query(condition, limit)
        elif workflow_id:
            condition = (
                Field("user_id")
                .equals(user_id)
                .and_(Field("workflow_id").equals(workflow_id))
                .and_(Field("id").greater_than(start_key or ""))
            )
            return cls.query(condition, limit)
        else:
            condition = (
                Field("user_id")
                .equals(user_id)
                .and_(Field("content_type").like((content_type or "") + "%"))
                .and_(Field("id").greater_than(start_key or ""))
            )
            return cls.query(condition, limit, reverse)

    @classmethod
    def get_children(cls, parent_id: str) -> Sequence["Asset"]:
        """
        Fetch all child assets for a given parent_id.
        """
        items, _ = cls.query(Field("parent_id").equals(parent_id))
        return items

    @classmethod
    def get_assets_recursive(cls, user_id: str, folder_id: str) -> Dict:
        """
        Fetch all assets recursively for a given folder_id.
        """
        def recursive_fetch(current_folder_id):
            assets, _ = cls.paginate(user_id=user_id, parent_id=current_folder_id, limit=10000)
            result = []
            for asset in assets:
                if asset.user_id != user_id:
                    continue
                
                asset_dict = asset.dict()
                if asset.content_type == "folder":
                    asset_dict["children"] = recursive_fetch(asset.id)
                result.append(asset_dict)
            
            return result

        folder = cls.find(user_id, folder_id)
        if not folder:
            log.warning(f"Folder {folder_id} not found for user {user_id}")
            return {"assets": []}

        folder_dict = folder.model_dump()
        folder_dict["children"] = recursive_fetch(folder_id)
        
        return {"assets": [folder_dict]}

File: packages/models/src/asset.ts
/**
 * Asset model – digital asset storage with folder hierarchy.
 *
 * Port of Python's `nodetool.models.asset`.
 */

import { eq, and, like, desc, isNull } from "drizzle-orm";
import { DBModel, createTimeOrderedUuid } from "./base-model.js";
import { getDb } from "./db.js";
import { assets } from "./schema/assets.js";

export class Asset extends DBModel {
  static override table = assets;

  declare id: string;
  declare user_id: string;
  declare parent_id: string | null;
  declare file_id: string | null;
  declare name: string;
  declare content_type: string;
  declare size: number | null;
  declare duration: number | null;
  declare metadata: Record<string, unknown> | null;
  declare workflow_id: string | null;
  declare node_id: string | null;
  declare job_id: string | null;
  declare created_at: string;
  declare updated_at: string;

  constructor(data: Record<string, unknown>) {
    super(data);
    const now = new Date().toISOString();
    this.id ??= createTimeOrderedUuid();
    this.name ??= "";
    this.content_type ??= "application/octet-stream";
    this.parent_id ??= null;
    this.file_id ??= null;
    this.size ??= null;
    this.duration ??= null;
    this.metadata ??= null;
    this.workflow_id ??= null;
    this.node_id ??= null;
    this.job_id ??= null;
    this.created_at ??= now;
    this.updated_at ??= now;
  }

  override beforeSave(): void {
    this.updated_at = new Date().toISOString();
  }

  // ── Computed properties ──────────────────────────────────────────

  get isFolder(): boolean {
    return this.content_type === "folder";
  }

  get fileExtension(): string {
    const dot = this.name.lastIndexOf(".");
    return dot >= 0 ? this.name.slice(dot + 1).toLowerCase() : "";
  }

  get hasThumbnail(): boolean {
    const type = this.content_type;
    return type.startsWith("image/") || type.startsWith("video/");
  }

  // ── Static queries ───────────────────────────────────────────────

  /** Find an asset by id, scoped to the user. */
  static async find(userId: string, assetId: string): Promise<Asset | null> {
    const asset = await Asset.get<Asset>(assetId);
    if (!asset || asset.user_id !== userId) return null;
    return asset;
  }

  /** List assets in a folder. */
  static async paginate(
    userId: string,
    opts: {
      parentId?: string | null;
      contentType?: string;
      workflowId?: string;
      nodeId?: string;
      jobId?: string;
      limit?: number;
    } = {}
  ): Promise<[Asset[], string]> {
    const {
      parentId,
      contentType,
      workflowId,
      nodeId,
      jobId,
      limit = 50
    } = opts;
    const db = getDb();

    const conditions = [eq(assets.user_id, userId)];
    if (parentId !== undefined) {
      if (parentId === null) {
        conditions.push(isNull(assets.parent_id));
      } else {
        conditions.push(eq(assets.parent_id, parentId));
      }
    }
    if (contentType) {
      conditions.push(eq(assets.content_type, contentType));
    }
    if (workflowId) {
      conditions.push(eq(assets.workflow_id, workflowId));
    }
    if (nodeId) {
      conditions.push(eq(assets.node_id, nodeId));
    }
    if (jobId) {
      conditions.push(eq(assets.job_id, jobId));
    }

    const rows = db
      .select()
      .from(assets)
      .where(and(...conditions))
      .orderBy(desc(assets.created_at))
      .limit(limit + 1)
      .all();

    const items = rows.map((r) => new Asset(r as Record<string, unknown>));
    if (items.length <= limit) return [items, ""];
    items.pop();
    const cursor = items[items.length - 1]?.id ?? "";
    return [items, cursor];
  }

  /** Get children of a folder. */
  static async getChildren(
    userId: string,
    parentId: string,
    limit = 100
  ): Promise<Asset[]> {
    const [assetList] = await Asset.paginate(userId, { parentId, limit });
    return assetList;
  }

  /**
   * Search assets globally across all folders for a user.
   * Uses LIKE on name field for substring matching.
   */
  static async searchAssetsGlobal(
    userId: string,
    query: string,
    opts: {
      contentType?: string;
      limit?: number;
    } = {}
  ): Promise<[Asset[], string, Array<Record<string, string>>]> {
    const { contentType, limit = 100 } = opts;
    // Escape LIKE special characters to prevent pattern injection
    const sanitized = query.trim().replace(/[%_\\]/g, "\\$&");
    const db = getDb();

    const conditions = [
      eq(assets.user_id, userId),
      like(assets.name, `%${sanitized}%`)
    ];
    if (contentType) {
      const sanitizedType = contentType.replace(/[%_\\]/g, "\\$&");
      conditions.push(like(assets.content_type, `${sanitizedType}%`));
    }

    const rows = db
      .select()
      .from(assets)
      .where(and(...conditions))
      .limit(limit)
      .all();

    const items = rows.map((r) => new Asset(r as Record<string, unknown>));
    const cursor = "";

    const pathInfo = await Asset.getAssetPathInfo(
      userId,
      items.map((a) => a.id)
    );

    const folderPaths: Array<Record<string, string>> = [];
    for (const asset of items) {
      if (pathInfo[asset.id]) {
        folderPaths.push(pathInfo[asset.id]);
      } else {
        folderPaths.push({
          folder_name: "Unknown",
          folder_path: "Unknown",
          folder_id: asset.parent_id ?? ""
        });
      }
    }

    return [items, cursor, folderPaths];
  }

  /**
   * Get folder path information for given asset IDs.
   */
  static async getAssetPathInfo(
    userId: string,
    assetIds: string[]
  ): Promise<Record<string, Record<string, string>>> {
    if (assetIds.length === 0) return {};

    const result: Record<string, Record<string, string>> = {};

    const assetMap = new Map<string, Asset>();
    for (const id of assetIds) {
      const asset = await Asset.find(userId, id);
      if (asset) assetMap.set(id, asset);
    }

    const parentCache = new Map<string, Asset>();

    for (const assetId of assetIds) {
      const asset = assetMap.get(assetId);
      if (!asset) continue;

      if (!asset.parent_id || asset.parent_id === userId) {
        result[assetId] = {
          folder_name: "Home",
          folder_path: "Home",
          folder_id: userId
        };
        continue;
      }

      const pathParts: string[] = [];
      const pathIds: string[] = [];
      let currentId: string | null = asset.parent_id;

      while (currentId && currentId !== userId) {
        let parent = parentCache.get(currentId);
        if (!parent) {
          parent = (await Asset.find(userId, currentId)) ?? undefined;
          if (parent) parentCache.set(currentId, parent);
        }
        if (!parent) break;
        pathParts.push(parent.name);
        pathIds.push(parent.id);
        currentId = parent.parent_id;
      }

      pathParts.push("Home");
      pathIds.push(userId);
      pathParts.reverse();
      pathIds.reverse();

      const immediateName = pathParts[pathParts.length - 1];
      const immediateId = pathIds[pathIds.length - 1];

      result[assetId] = {
        folder_name: immediateName,
        folder_path: pathParts.join(" / "),
        folder_id: immediateId
      };
    }

    return result;
  }

  /**
   * Recursively fetch all assets within a folder.
   */
  static async getAssetsRecursive(
    userId: string,
    folderId: string
  ): Promise<{ assets: Record<string, unknown>[] }> {
    const folder = await Asset.find(userId, folderId);
    if (!folder) return { assets: [] };

    async function recursiveFetch(
      currentFolderId: string
    ): Promise<Record<string, unknown>[]> {
      const [assetList] = await Asset.paginate(userId, {
        parentId: currentFolderId,
        limit: 10000
      });
      const result: Record<string, unknown>[] = [];
      for (const asset of assetList) {
        const dict = asset.toRow();
        if (asset.content_type === "folder") {
          dict.children = await recursiveFetch(asset.id);
        }
        result.push(dict);
      }
      return result;
    }

    const folderDict = folder.toRow();
    folderDict.children = await recursiveFetch(folderId);

    return { assets: [folderDict] };
  }
}
